### PR TITLE
FIX undefined index error in CMS

### PR DESCRIPTION
**Pull request by NightJar**
With the CMS 4.12 update functionality was altered to utilise an Extension to obtain the CMS Edit link for a page, rather than having SiteTree do it internally. Unfortunately the default return case for `extend` (see Extensible) is an _empty_ array. This leaves code potentially referencing an array offset that doesn't exist ([0]). PHP 8 is less forgiving than its predecessors on this kind of behaviour. We should check that the responses from extensions exist before trying to reference them.

### CODE REVIEW

File: code/Model/SiteTree.php [hunk 1]
```diff
@@ -739,13 +739,16 @@ public function getAbsoluteLiveLink($includeStageEqualsLive = true)
...
+     *
      * @return string
      */
     public function CMSEditLink()
     {
-        // This method has to be implemented here to satisfy the CMSPreviewable interface.
-        // See the actual implementation in CMSEditLinkExtension.
-        return $this->extend('CMSEditLink')[0];
+        $extensionsResult = $this->extend('CMSEditLink');
+        return count($extensionsResult) ? $extensionsResult[0] : '';
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php]:
```suggestion
        return $extensionsResult[0] ?? '';
```
This is just a little clearer - as we don't really care about the number of items in the array so much as we care that there is a value in the `[0]` index.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Thank you, this is great.
Just a small change.

File: code/Model/SiteTree.php (hunk 1 above)
Comment (NightJar) [reply code/Model/SiteTree.php]: `extend` _always_ returns an `array`, never `null`.
***
Wait sorry, didn't see the dereference in the suggestion.
Does PHP not error on that still, as it's accessing an undefined index? Or does `??` squash errors?

Comment (ChloeSartorelli) [reply code/Model/SiteTree.php]: `??` squashes the error - basically says "If there's nothing there, use the next value"
It's called "null coalescing operator" but it actually does an `isset()` check.
https://www.php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op gives this as an explicitly recommended use case.

Comment (NightJar) [reply code/Model/SiteTree.php]: Neat thanks, I knew it was null coalescing, but always forget if it also handles the "null pointer" reference too.

File: code/Model/SiteTree.php [hunk 2]
```diff
@@ -739,13 +739,15 @@ public function getAbsoluteLiveLink($includeStageEqualsLive = true)
...
+     * @see SilverStripe\Admin\CMSEditLinkExtension
+     *
      * @return string
      */
     public function CMSEditLink()
     {
-        // This method has to be implemented here to satisfy the CMSPreviewable interface.
-        // See the actual implementation in CMSEditLinkExtension.
-        return $this->extend('CMSEditLink')[0];
+        return $this->extend('CMSEditLink') ?? '';
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php]:
```suggestion
        return $this->extend('CMSEditLink')[0] ?? '';
```
Still need to get the 0th result

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: code/Model/SiteTree.php (hunk 2 above)
Comment (NightJar) [reply code/Model/SiteTree.php]: oops haha 😅

**Comment by NightJar** [comment]
Is there anything left to do here?
Can you rerun the behat task please? I feel like this failure is due to typical behat test flakeyness, not this change.

**Review by GuySartorelli — APPROVED** [review]
Sorry, I got distracted and lost sight of this. Thank you for submitting it!

**Comment by NightJar** [comment]
Thanks for restarting the automated test run, it's a relief to see it all pass :)

**Pull request merged**